jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/PerProcessInstanceRuntimeManager.java

```diff
@@ -191,15 +194,35 @@ public void signalEvent(String type, Object event) {
         
         // process currently active runtime engines
         Map<Object, RuntimeEngine> currentlyActive = local.get();
+
         if (currentlyActive != null && !currentlyActive.isEmpty()) {
-            RuntimeEngine[] activeEngines = currentlyActive.values().toArray(new RuntimeEngine[currentlyActive.size()]);
-            for (RuntimeEngine engine : activeEngines) {
-                Context<?> context = ((RuntimeEngineImpl) engine).getContext();
+            @SuppressWarnings("unchecked")
+            Entry<Object, RuntimeEngine> activeEngines[] = currentlyActive.entrySet()
+                    .toArray(new Entry[currentlyActive.size()]);
+            Set<Object> enginesToDelete = new HashSet<>();
+            for (Entry<Object, RuntimeEngine> engine : activeEngines) {
+                RuntimeEngineImpl engineImpl = (RuntimeEngineImpl) engine.getValue();
+                if (engineImpl.isDisposed() || engineImpl.isInvalid()) {
+                    enginesToDelete.add(engine.getKey());
+                    continue;
+                }
+                Context<?> context = engineImpl.getContext();
                 if (context != null && context instanceof ProcessInstanceIdContext 
                         && ((ProcessInstanceIdContext) context).getContextId() != null) {
-                    engine.getKieSession().signalEvent(type, event, ((ProcessInstanceIdContext) context).getContextId());
+                    try {
+                        engineImpl.getKieSession().signalEvent(type, event,
+                                ((ProcessInstanceIdContext) context).getContextId());
+                    } catch (org.drools.persistence.api.SessionNotFoundException ex) {
+                        logger.warn(
+                                "Signal event cannot proceed because of session not found exception {} for engine {}",
+                                ex.getMessage(), engineImpl.getKieSessionId());
+                        enginesToDelete.add(engine.getKey());
+                    }
                 }
             }
+            if (!enginesToDelete.isEmpty()) {
+                currentlyActive.keySet().removeAll(enginesToDelete);
```

Comment:
you could log the enginesToDelete for tracking (debug level if you don't want to make it verbose)